contracts/Staking.sol

```diff
@@ -334,40 +356,31 @@ contract Staking is Autopetrified, ERCStaking, ERCStakingHistory, IStakingLockin
      * @param _blockNumber Block number at which we are requesting
      * @return The amount of tokens staked at the given block number
      */
-    function totalStakedAt(uint256 _blockNumber) external view returns (uint256) {
+    function totalStakedAt(uint256 _blockNumber) external view isInitialized returns (uint256) {
         return totalStakedHistory.get(_blockNumber);
     }
 
-    /* Public functions */
-
-    /**
-     * @notice Unlock `_accountAddress`'s lock #`_lockId` so locked tokens can be unstaked again
-     * @param _accountAddress Owner of locked tokens
-     * @param _lockId Id of the lock for the given account
-     */
-    function unlock(address _accountAddress, uint256 _lockId) public {
-        // only manager and owner (if manager allows) can unlock
-        require(canUnlock(_accountAddress, _lockId), ERROR_CAN_NOT_UNLOCK);
-
-        _unlock(_accountAddress, _lockId);
-    }
-
     /**
      * @notice Get the staked but unlocked amount of tokens by `_accountAddress`
      * @param _accountAddress Owner of the staked but unlocked balance
      * @return Amount of tokens staked but not locked by given account
      */
-    function unlockedBalanceOf(address _accountAddress) public view returns (uint256) {
-        uint256 unlockedTokens = totalStakedFor(_accountAddress);
-
-        Account storage account = accounts[_accountAddress];
-        for (uint256 i = 0; i < account.activeLockIds.length; i++) {
-            unlockedTokens = unlockedTokens.sub(account.locks[account.activeLockIds[i]].amount);
-        }
+    function unlockedBalanceOf(address _accountAddress) external view isInitialized returns (uint256) {
+        return _unlockedBalanceOf(_accountAddress);
+    }
 
-        return unlockedTokens;
+    /**
+     * @notice Check if `_accountAddress`'s by `_lockManager` can be unlocked
+     * @param _accountAddress Owner of lock
+     * @param _lockManager Manager of the lock for the given account
+     * @return Whether given lock of given account can be unlocked
+     */
+    function canUnlock(address _accountAddress, address _lockManager, uint256 _amount) external view isInitialized returns (bool) {
+        return _canUnlock(_accountAddress, _lockManager, _amount);
```

Comment:
This will always return false unless the sender is the manager or someone the manager allowed previously, right? What do you think about parameterizing the sender for the internal function?

Reply:
Not exactly, it returns false unless the sender is the manager, or the owner under certain conditions set by the manager. But always the manager or the owner, nobody else. That’s why I think it’s better to keep using `msg.sender`, to avoid mis-using the function.